File: Drift/EventViewController.m
//
//  EventViewController.m
//  Drift
//
//  Created by Mikael Hultgren on 5/31/12.
//  Copyright (c) 2012 Artsoftheinsane. All rights reserved.
//

#import "EventViewController.h"

#import "Event.h"
#import "Tag.h"
#import "TagsTableViewController.h"
#import <QuartzCore/QuartzCore.h>
#import "State.h"
#import "NSDate+Utilities.h"
#import "CAAnimation+Blocks.h"

static void *EventViewControllerContext = &EventViewControllerContext;

@interface EventViewController () <SlideUpTransitioningDelegate, DismissProtocol>

@property (nonatomic) NSArray *shortStandaloneMonthSymbols;
@property (nonatomic) Event *selectedEvent;

//@property (nonatomic) SlideUpTransitioning *slideUpTransitioning;

@end

@implementation EventViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.shortStandaloneMonthSymbols = [[NSDateFormatter new] shortStandaloneMonthSymbols];

//    self.slideUpTransitioning = [SlideUpTransitioning new];
//    self.slideUpTransitioning.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    [self.eventTimerControl addObserver:self
                             forKeyPath:NSStringFromSelector(@selector(startDate))
                                options:(NSKeyValueObservingOptionNew)
                                context:EventViewControllerContext];

    [self.eventTimerControl addObserver:self
                             forKeyPath:NSStringFromSelector(@selector(nowDate))
                                options:(NSKeyValueObservingOptionNew)
                                context:EventViewControllerContext];

    [self.eventTimerControl addObserver:self
                             forKeyPath:NSStringFromSelector(@selector(transforming))
                                options:(NSKeyValueObservingOptionNew)
                                context:EventViewControllerContext];

    NSString *guid = [State instance].selectedEventGUID;
    if (guid) {
        self.selectedEvent = [Event MR_findFirstByAttribute:@"guid"
                                             withValue:guid];
    } else {
        self.selectedEvent = nil;
    }

    NSString *name = nil;
    if (self.selectedEvent) {
        [self.eventTimerControl initWithStartDate:self.selectedEvent.startDate
                                      andStopDate:self.selectedEvent.stopDate];

        if (self.selectedEvent.isActive) {
            [self.toggleStartStopButton setTitle:@"STOP"
                                        forState:UIControlStateNormal];
            [self animateStartEvent];
        } else {
            [self.toggleStartStopButton setTitle:@"START"
                                        forState:UIControlStateNormal];
            [self animateStopEvent];
        }

        name = self.selectedEvent.inTag.name;
    } else {
        [State instance].selectedEventGUID = nil;

        [self reset];
        [self.eventTimerControl reset];
    }

    NSAttributedString *attributeString = nil;
    if (name) {
        attributeString = [[NSAttributedString alloc] initWithString:name
                                                          attributes:@{NSFontAttributeName:[UIFont fontWithName:@"Helvetica Neue" size:14]}];
    } else {
        attributeString = [[NSAttributedString alloc] initWithString:@"\uf02b"
                                                          attributes:@{NSFontAttributeName:[UIFont fontWithName:@"FontAwesome" size:20]}];
    }

    [self.tag setAttributedTitle:attributeString
                        forState:UIControlStateNormal];

}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

//    if (![self.view.window.gestureRecognizers containsObject:self.slideUpTransitioning.gestureRecogniser]) {
//        [self.view.window addGestureRecognizer:self.slideUpTransitioning.gestureRecogniser];
//    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];

    [self.eventTimerControl paus];

    [self.eventTimerControl removeObserver:self
                                forKeyPath:NSStringFromSelector(@selector(startDate))];
    [self.eventTimerControl removeObserver:self
                                forKeyPath:NSStringFromSelector(@selector(nowDate))];
    [self.eventTimerControl removeObserver:self
                                forKeyPath:NSStringFromSelector(@selector(transforming))];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"segueToTagsFromEvent"]) {
        TagsTableViewController *controller = (TagsTableViewController *)[segue destinationViewController];
        __weak __typeof__(self) _self = self;
        [controller setDidDismissHandler: ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [_self dismissViewControllerAnimated:YES completion:nil];
            });
        }];

        controller.eventGUID = self.selectedEvent.guid;
    } else if ([segue.identifier isEqualToString:@"segueToEventsFromEvent"]) {
        EventsViewController *controller = (EventsViewController *)[segue destinationViewController];
//        controller.transitioningDelegate = self.slideUpTransitioning;
        controller.dismissDelegate = self;
    }
}


#pragma mark -
#pragma mark DismissProtocol

- (void)didDismiss {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark -
#pragma mark SlideUpTransitioningDelegate

- (void)proceedToNextViewController {
    [self performSegueWithIdentifier:@"segueToEventsFromEvent"
                              sender:self];
}

#pragma mark -
#pragma mark Public methods

- (IBAction)showTags:(id)sender {
    if (self.selectedEvent) {
        [self animateButton:sender];

        [self performSegueWithIdentifier:@"segueToTagsFromEvent"
                                  sender:self];
    }
}

- (IBAction)toggleEventTouchUpInside:(id)sender forEvent:(UIEvent *)event {
    if (self.selectedEvent.isActive) {
        [self.eventTimerControl stop];
        self.selectedEvent.stopDate = self.eventTimerControl.nowDate;

        [self.toggleStartStopButton setTitle:@"START"
                                    forState:UIControlStateNormal];
        [self animateStopEvent];
    } else {
        [self reset];

        self.selectedEvent = [Event MR_createEntity];
        self.selectedEvent.startDate = [NSDate date];

        [State instance].selectedEventGUID = self.selectedEvent.guid;

        [self.eventTimerControl initWithStartDate:self.selectedEvent.startDate
                                      andStopDate:self.selectedEvent.stopDate];

        [self.toggleStartStopButton setTitle:@"STOP"
                                    forState:UIControlStateNormal];
        [self animateStartEvent];

        NSAttributedString *attributeString = [[NSAttributedString alloc] initWithString:@"\uf02b"
                                                                              attributes:@{NSFontAttributeName:[UIFont fontWithName:@"FontAwesome" size:20]}];
        [self.tag setAttributedTitle:attributeString forState:UIControlStateNormal];
    }

    [self animateButton:sender];

    [[NSManagedObjectContext MR_defaultContext] MR_saveToPersistentStoreWithCompletion:nil];
}

#pragma mark -
#pragma mark Private methods

- (void)animateButton:(UIButton *)button {
    CGRect       pathFrame = CGRectMake(-CGRectGetMidY(button.bounds), -CGRectGetMidY(button.bounds), button.bounds.size.height, button.bounds.size.height);
    UIBezierPath *path     = [UIBezierPath bezierPathWithRoundedRect:pathFrame cornerRadius:pathFrame.size.height / 2];

    CGPoint shapePosition = [self.view convertPoint:button.center fromView:button.superview];

    CAShapeLayer *circleShape = [CAShapeLayer layer];
    circleShape.path        = path.CGPath;
    circleShape.position    = shapePosition;
    circleShape.fillColor   = [UIColor clearColor].CGColor;
    circleShape.opacity     = 0;
    circleShape.strokeColor = button.titleLabel.textColor.CGColor;
    circleShape.lineWidth   = 2;

    [self.view.layer addSublayer:circleShape];

    CABasicAnimation *scaleAnimation = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    scaleAnimation.fromValue = [NSValue valueWithCATransform3D:CATransform3DIdentity];
    scaleAnimation.toValue   = [NSValue valueWithCATransform3D:CATransform3DMakeScale(3, 3, 1)];

    CABasicAnimation *alphaAnimation = [CABasicAnimation animationWithKeyPath:@"opacity"];
    alphaAnimation.fromValue = @1;
    alphaAnimation.toValue   = @0;

    CAAnimationGroup *animation = [CAAnimationGroup animation];
    animation.animations     = @[scaleAnimation, alphaAnimation];
    animation.duration       = 0.5f;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    [animation setCompletion: ^(BOOL finished) {
        [circleShape removeFromSuperlayer];
    }];

    [circleShape addAnimation:animation forKey:nil];
}

- (void)reset {
    [self.toggleStartStopButton setTitle:@"START"
                                forState:UIControlStateNormal];

    self.eventStartTime.text    = @"";
    self.eventStartDay.text     = @"";
    self.eventStartYear.text    = @"";
    self.eventStartMonth.text   = @"";

    self.eventTimeHours.text    = @"00";
    self.eventTimeMinutes.text  = @"00";

    self.eventStopTime.text     = @"";
    self.eventStopDay.text      = @"";
    self.eventStopYear.text     = @"";
    self.eventStopMonth.text    = @"";

    self.eventStartDay.alpha    = 1;
    self.eventStartMonth.alpha  = 1;
    self.eventStartTime.alpha   = 1;
    self.eventStartYear.alpha   = 1;

    self.eventStopDay.alpha     = 1;
    self.eventStopMonth.alpha   = 1;
    self.eventStopTime.alpha    = 1;
    self.eventStopYear.alpha    = 1;

    self.eventTimeHours.alpha   = 1;
    self.eventTimeMinutes.alpha = 1;
}

- (void)updateStartLabelWithDate:(NSDate *)date {
    if (date) {
        static NSUInteger unitFlags  = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitWeekday | NSCalendarUnitDay | NSCalendarUnitHour | NSCalendarUnitMinute;
        NSDateComponents *components = [[NSDate calendar] components:unitFlags
                                                            fromDate:date];

        self.eventStartTime.text  = [NSString stringWithFormat:@"%02ld:%02ld", (long)components.hour, (long)components.minute];
        self.eventStartDay.text   = [NSString stringWithFormat:@"%02ld", (long)components.day];
        self.eventStartYear.text  = [NSString stringWithFormat:@"%04ld", (long)components.year];
        self.eventStartMonth.text = [self.shortStandaloneMonthSymbols objectAtIndex:(NSUInteger)(components.month - 1)];
    }
}

- (void)updateEventTimeFromDate:(NSDate *)fromDate toDate:(NSDate *)toDate {
    if (fromDate && toDate) {
        static NSUInteger unitFlags  = NSCalendarUnitHour | NSCalendarUnitMinute;
        NSDateComponents  *components = [[NSDate calendar] components:unitFlags
                                                             fromDate:fromDate
                                                               toDate:toDate options:0];

        NSInteger hour = ABS(components.hour);
        NSInteger minute = ABS(components.minute);

        NSString *eventTimeHours = [NSString stringWithFormat:@"%02ld", (long)hour];
        if (components.hour < 0 || components.minute < 0) {
            eventTimeHours = [NSString stringWithFormat:@"-%@", eventTimeHours];
        }

        self.eventTimeHours.text   = eventTimeHours;
        self.eventTimeMinutes.text = [NSString stringWithFormat:@"%02ld", (long)minute];
    }
}

- (void)updateStopLabelWithDate:(NSDate *)date {
    if (date) {
        static NSUInteger unitFlags  = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitWeekday | NSCalendarUnitDay | NSCalendarUnitHour | NSCalendarUnitMinute;
        NSDateComponents *components = [[NSDate calendar] components:unitFlags
                                                            fromDate:date];

        self.eventStopTime.text  = [NSString stringWithFormat:@"%02ld:%02ld", (long)components.hour, (long)components.minute];
        self.eventStopDay.text   = [NSString stringWithFormat:@"%02ld", (long)components.day];
        self.eventStopYear.text  = [NSString stringWithFormat:@"%04ld", (long)components.year];
        self.eventStopMonth.text = [self.shortStandaloneMonthSymbols objectAtIndex:(NSUInteger)(components.month - 1)];
    }
}

- (void)animateStartEvent {
    [UIView animateWithDuration:0.3
                          delay:0.0
                        options:UIViewAnimationOptionCurveEaseIn
                     animations:^{
        self.eventStartTime.alpha = 1;
        self.eventStartDay.alpha = 1;
        self.eventStartMonth.alpha = 1;
        self.eventStartYear.alpha = 1;

        self.eventStopTime.alpha = 0.2f;
        self.eventStopDay.alpha = 0.2f;
        self.eventStopMonth.alpha = 1;
        self.eventStopYear.alpha = 1;
    } completion:nil];
}

- (void)animateStopEvent {
    [UIView animateWithDuration:0.3
                          delay:0.0
                        options:UIViewAnimationOptionCurveEaseIn
                     animations:^{
        self.eventStartTime.alpha = 0.2f;
        self.eventStartDay.alpha = 0.2f;
        self.eventStartMonth.alpha = 1;
        self.eventStartYear.alpha = 1;

        self.eventStopTime.alpha = 1;
        self.eventStopDay.alpha = 1;
        self.eventStopMonth.alpha = 1;
        self.eventStopYear.alpha = 1;
    } completion:nil];
}

- (void)animateEventTransforming:(EventTimerTransformingEnum)eventTimerTransformingEnum {
    [UIView animateWithDuration:0.3
                          delay:0.0
                        options:UIViewAnimationOptionCurveEaseIn
                     animations:
     ^{
         CGFloat eventStartAlpha = 1, eventStopAlpha = 1, eventTimeAlpha = 1, eventStartMonthYearAlpha = 1, eventStopMonthYearAlpha = 1;
         switch (eventTimerTransformingEnum) {
             case EventTimerStartDateTransformingStart:
                 eventStartAlpha = 1;
                 eventStartMonthYearAlpha = 1;

                 eventStopAlpha = 0.2f;
                 eventStopMonthYearAlpha = 0.2f;

                 eventTimeAlpha = 0.2f;
                 break;
             case EventTimerStartDateTransformingStop:
                 eventStartAlpha = self.selectedEvent.isActive ? 1 : 0.2f;
                 eventStartMonthYearAlpha = 1;

                 eventStopAlpha = self.selectedEvent.isActive ? 0.2f : 1;
                 eventStopMonthYearAlpha = 1;

                 eventTimeAlpha = 1;
                 break;
             case EventTimerNowDateTransformingStart:
                 eventStartAlpha = 0.2f;
                 eventStartMonthYearAlpha = 0.2f;

                 eventStopAlpha = 1;
                 eventStopMonthYearAlpha = 1;

                 eventTimeAlpha = 0.2f;

                 break;
             case EventTimerNowDateTransformingStop:
                 eventStartAlpha = self.selectedEvent.isActive ? 1 : 0.2f;
                 eventStartMonthYearAlpha = 1;

                 eventStopAlpha = self.selectedEvent.isActive ? 0.2f : 1;
                 eventStopMonthYearAlpha = 1;

                 eventTimeAlpha = 1;
                 break;
             default:
                 break;
         }

         self.eventStartDay.alpha = eventStartAlpha;
         self.eventStartMonth.alpha = eventStartMonthYearAlpha;
         self.eventStartTime.alpha = eventStartAlpha;
         self.eventStartYear.alpha = eventStartMonthYearAlpha;

         self.eventStopDay.alpha = eventStopAlpha;
         self.eventStopMonth.alpha = eventStopMonthYearAlpha;
         self.eventStopTime.alpha = eventStopAlpha;
         self.eventStopYear.alpha = eventStopMonthYearAlpha;

         self.eventTimeHours.alpha = eventTimeAlpha;
         self.eventTimeMinutes.alpha = eventTimeAlpha;
     } completion:nil];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    if (context == EventViewControllerContext) {
        int changeKindKey = [[change objectForKey:NSKeyValueChangeKindKey] intValue];

        if ([keyPath isEqualToString:NSStringFromSelector(@selector(startDate))]) {
            if (changeKindKey == NSKeyValueChangeSetting) {
                id newValue = [change objectForKey:NSKeyValueChangeNewKey];

                NSDate *toDate = self.eventTimerControl.nowDate;
                [self updateStartLabelWithDate:newValue];
                [self updateEventTimeFromDate:newValue toDate:toDate];
            }
        } else if ([keyPath isEqualToString:NSStringFromSelector(@selector(nowDate))]) {
            if (changeKindKey == NSKeyValueChangeSetting) {
                id newValue = [change objectForKey:NSKeyValueChangeNewKey];

                NSDate *fromDate = self.eventTimerControl.startDate;
                [self updateStopLabelWithDate:newValue];
                [self updateEventTimeFromDate:fromDate toDate:newValue];
            }
        } else if ([keyPath isEqualToString:NSStringFromSelector(@selector(transforming))]) {
            if (changeKindKey == NSKeyValueChangeSetting) {
                id newValue = [change objectForKey:NSKeyValueChangeNewKey];

                EventTimerTransformingEnum transforming = [newValue integerValue];

                if (transforming != EventTimerNotTransforming) {
                    [self animateEventTransforming:transforming];
                }

                switch (transforming) {
                    case EventTimerNowDateTransformingStart:
                    case EventTimerStartDateTransformingStart:
//                        self.slideUpTransitioning.gestureRecogniser.enabled = NO;

                        break;
                    case EventTimerNowDateTransformingStop:
//                        self.slideUpTransitioning.gestureRecogniser.enabled = YES;

                        self.selectedEvent.stopDate = self.eventTimerControl.nowDate;
                        [[NSManagedObjectContext MR_defaultContext] MR_saveToPersistentStoreWithCompletion:nil];

                        break;
                    case EventTimerStartDateTransformingStop:
//                        self.slideUpTransitioning.gestureRecogniser.enabled = YES;

                        self.selectedEvent.startDate = self.eventTimerControl.startDate;
                        [[NSManagedObjectContext MR_defaultContext] MR_saveToPersistentStoreWithCompletion:nil];

                        break;

                    default:
                        break;
                }
            }
        }
    }
}

@end

File: Drift/EventViewController.swift
import UIKit
import CoreData

class EventViewController: UIViewController, EventTimerControlDelegate, CoreDataInjected {
    // MARK: IBOutlet
    @IBOutlet weak var eventTimerControl: EventTimerControl?
    @IBOutlet weak var toggleStartStopButton: UIButton?

    @IBOutlet weak var eventStartTime: UILabel?
    @IBOutlet weak var eventStartDay: UILabel?
    @IBOutlet weak var eventStartMonth: UILabel?
    @IBOutlet weak var eventStartYear: UILabel?

    @IBOutlet weak var eventTimeHours: UILabel?
    @IBOutlet weak var eventTimeMinutes: UILabel?

    @IBOutlet weak var eventStopTime: UILabel?
    @IBOutlet weak var eventStopDay: UILabel?
    @IBOutlet weak var eventStopMonth: UILabel?
    @IBOutlet weak var eventStopYear: UILabel?

    @IBOutlet weak var tag: UIButton?

    // MARK: Private properties
    private var selectedEventID: URL?

    private let state = State()
    private var transitionOperator: TransitionOperator?

    private let calendar = Calendar.autoupdatingCurrent
    private let shortStandaloneMonthSymbols: NSArray = DateFormatter().shortStandaloneMonthSymbols

    override func viewDidLoad() {
        super.viewDidLoad()

        self.transitionOperator = TransitionOperator(viewController: self)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        eventTimerControl?.delegate = self

        var tagName: String? = nil
        if let url = state.selectedEventID {
			let result: Result<Event, FetchError> = fetch(url: url, inContext: persistentContainer.viewContext)
			do {
				let event = try result.dematerialize()
				selectedEventID = url

				tagName = event.inTag?.name
				eventTimerControl?.initWithStart(event.startDate, andStop: event.stopDate)
				
				if let _ = event.stopDate {
					toggleStartStopButton?.setTitle("START", for: UIControlState())
					animateStopEvent()
				} else {
					toggleStartStopButton?.setTitle("STOP", for: UIControlState())
					animateStartEvent()
				}
			} catch {
				// TODO: Error handling
			}
        } else {
            selectedEventID = nil
        }

        if let tagName = tagName,
            let font = UIFont(name: "Helvetica Neue", size: 14) {
                let attriString = AttributedString(string:tagName, attributes:
                    [NSFontAttributeName: font])

                tag?.setAttributedTitle(attriString, for: UIControlState())
        } else if let font = UIFont(name: "FontAwesome", size: 20) {
            let attriString = AttributedString(string:"\u{f02b}", attributes:
                [NSFontAttributeName: font])

            tag?.setAttributedTitle(attriString, for: UIControlState())
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

		eventTimerControl?.delegate = nil
		eventTimerControl?.stop()
    }

	override func prepare(for segue: UIStoryboardSegue, sender: AnyObject?) {
		if segue.identifier == "segueToTagsFromEvent",
			let controller = segue.destinationViewController as? TagsViewController,
            let id = selectedEventID {
				controller.eventID = id
		}
	}

    private func updateStartLabelWithDate(_ date: Date) {
        let unitFlags: Calendar.Unit = [.year, .month, .day, .hour, .minute]

        let components: DateComponents = calendar.components(unitFlags, from: date)

        eventStartTime?.text  = String(format: "%02ld:%02ld", components.hour!, components.minute!)
        eventStartDay?.text  = String(format: "%02ld", components.day!)
        eventStartYear?.text  = String(format: "%04ld", components.year!)
        let index = components.month! - 1
        if let month = shortStandaloneMonthSymbols.object(at: index) as? String {
            eventStartMonth?.text  = month
        }
    }

    private func updateEventTimeFromDate(_ fromDate: Date, toDate: Date) {
        let unitFlags: Calendar.Unit = [.hour, .minute]
        let components: DateComponents = calendar.components(unitFlags, from: fromDate, to: toDate, options: [])

        let hour: Int = abs(components.hour!)
        let minute: Int = abs(components.minute!)

        var timeHours: String = String(format:"%02ld", hour)
        if components.hour < 0 || components.minute < 0 {
            timeHours = String(format:"-%@", timeHours)
        }

        eventTimeHours?.text   = timeHours
        eventTimeMinutes?.text = String(format:"%02ld", minute)
    }

    private func updateStopLabelWithDate(_ date: Date) {
        let unitFlags: Calendar.Unit = [.year, .month, .day, .hour, .minute]

        let components: DateComponents = calendar.components(unitFlags, from: date)

        eventStopTime?.text  = String(format: "%02ld:%02ld", components.hour!, components.minute!)
        eventStopDay?.text  = String(format: "%02ld", components.day!)
        eventStopYear?.text  = String(format: "%04ld", components.year!)
        let index = components.month! - 1
        if let month = shortStandaloneMonthSymbols.object(at: index) as? String {
            eventStopMonth?.text  = month
        }
    }

    private func animateStartEvent() {
        UIView.animate(withDuration: 0.3, delay: 0.3, options: .curveEaseIn, animations: { () -> Void in
            self.eventStartTime?.alpha = 1
            self.eventStartDay?.alpha = 1
            self.eventStartMonth?.alpha = 1
            self.eventStartYear?.alpha = 1

            self.eventStopTime?.alpha = 0.2
            self.eventStopDay?.alpha = 0.2
            self.eventStopMonth?.alpha = 1
            self.eventStopYear?.alpha = 1
        }, completion: nil)
    }

    private func animateStopEvent() {
        UIView.animate(withDuration: 0.3, delay: 0.3, options: .curveEaseIn, animations: { () -> Void in
            self.eventStartTime?.alpha = 0.2
            self.eventStartDay?.alpha = 0.2
            self.eventStartMonth?.alpha = 1
            self.eventStartYear?.alpha = 1

            self.eventStopTime?.alpha = 1
            self.eventStopDay?.alpha = 1
            self.eventStopMonth?.alpha = 1
            self.eventStopYear?.alpha = 1
            }, completion: nil)
    }

    private func animateEventTransforming(_ eventTimerTransformingEnum: EventTimerTransformingEnum) {
//        UIView.animate(withDuration: 0.3, delay: 0.3, options: .curveEaseIn, animations: { () -> Void in
//            var eventStartAlpha: CGFloat = 1
//            var eventStopAlpha: CGFloat = 1
//            var eventTimeAlpha: CGFloat = 1
//            var eventStartMonthYearAlpha: CGFloat = 1
//            var eventStopMonthYearAlpha: CGFloat = 1
//
//            switch eventTimerTransformingEnum {
//            case .startDateTransformingStart:
//                eventStartAlpha = 1
//                eventStartMonthYearAlpha = 1
//
//                eventStopAlpha = 0.2
//                eventStopMonthYearAlpha = 0.2
//
//                eventTimeAlpha = 0.2
//            case .startDateTransformingStop:
//                if let _ = self.selectedEvent(self.selectedEventGuid)?.stopDate {
//                    eventStartAlpha = 0.2
//                    eventStopAlpha = 1
//                } else {
//                    eventStartAlpha = 1
//                    eventStopAlpha = 0.2
//                }
//
//                eventStartMonthYearAlpha = 1
//                eventStopMonthYearAlpha = 1
//
//                eventTimeAlpha = 1
//            case .nowDateTransformingStart:
//                eventStartAlpha = 0.2
//                eventStartMonthYearAlpha = 0.2
//
//                eventStopAlpha = 1
//                eventStopMonthYearAlpha = 1
//
//                eventTimeAlpha = 0.2
//            case .nowDateTransformingStop:
//                if let _ = self.selectedEvent(self.selectedEventGuid)?.stopDate {
//                    eventStartAlpha = 0.2
//                    eventStopAlpha = 1
//                } else {
//                    eventStartAlpha = 1
//                    eventStopAlpha = 0.2
//                }
//
//                eventStartMonthYearAlpha = 1
//                eventStopMonthYearAlpha = 1
//
//                eventTimeAlpha = 1
//            default:
//                break
//            }
//
//            self.eventStartDay?.alpha = eventStartAlpha
//            self.eventStartMonth?.alpha = eventStartMonthYearAlpha
//            self.eventStartTime?.alpha = eventStartAlpha
//            self.eventStartYear?.alpha = eventStartMonthYearAlpha
//
//            self.eventStopDay?.alpha = eventStopAlpha
//            self.eventStopMonth?.alpha = eventStopMonthYearAlpha
//            self.eventStopTime?.alpha = eventStopAlpha
//            self.eventStopYear?.alpha = eventStopMonthYearAlpha
//
//            self.eventTimeHours?.alpha = eventTimeAlpha
//            self.eventTimeMinutes?.alpha = eventTimeAlpha
//
//            }, completion: nil)
    }

	@IBAction func prepareForUnwind(_ sender: UIStoryboardSegue) {
		DLog()
	}

    @IBAction func showTags(_ sender: UIButton) {
        if selectedEventID != nil {
            navigationController?.delegate = nil
            performSegue(withIdentifier: "segueToTagsFromEvent", sender: self)
        }
    }

    @IBAction func toggleEventTouchUpInside(_ sender: UIButton) {
        if let id = selectedEventID {
			let result: Result<Event, FetchError> = fetch(url: id, inContext: persistentContainer.viewContext)

			guard let event = result.value() else {
				// For some reason there was no event
				// attached to this objectid, not at all what
				// was exspected
				// TODO: Error handling
				fatalError()
			}

            if let _ = event.stopDate {
				// Event is stoped, so start a new
				let event = Event(inContext: persistentContainer.viewContext)
				event.startDate = Date()

				do {
					try save(context: persistentContainer.viewContext)
				} catch {
					// TODO: Error handling
					print("*** ERROR: [\(#line)] \(#function) \(error)")
				}

                selectedEventID = event.objectID.uriRepresentation()
				state.selectedEventID = selectedEventID

				eventTimerControl?.initWithStart(event.startDate, andStop: event.stopDate)

				toggleStartStopButton?.setTitle("STOP", for: UIControlState())

				animateStartEvent()

				if let font = UIFont(name: "FontAwesome", size: 20) {
					let attriString = AttributedString(string:"\u{f02b}", attributes:
						[NSFontAttributeName: font])

					tag?.setAttributedTitle(attriString, for: UIControlState())
				}
            } else {
                // Event is started
                eventTimerControl?.stop()
                event.stopDate = eventTimerControl?.nowDate

                toggleStartStopButton?.setTitle("START", for: UIControlState())
                animateStopEvent()
            }
        } else {
			// No event exists, start a new
			// Event is stoped, so start a new
			let event = Event(inContext: persistentContainer.viewContext)
			event.startDate = Date()

			// NOTE: Only needed to get permanent object id
			do {
				try save(context: persistentContainer.viewContext)
			} catch {
				// TODO: Error handling
				print("*** ERROR: [\(#line)] \(#function) \(error)")
			}

            selectedEventID = event.objectID.uriRepresentation()
            state.selectedEventID = selectedEventID

			eventTimerControl?.initWithStart(event.startDate, andStop: event.stopDate)

			toggleStartStopButton?.setTitle("STOP", for: UIControlState())

			animateStartEvent()

			if let font = UIFont(name: "FontAwesome", size: 20) {
				let attriString = AttributedString(string:"\u{f02b}", attributes:
					[NSFontAttributeName: font])

				tag?.setAttributedTitle(attriString, for: UIControlState())
			}
		}

		do {
			try save(context: persistentContainer.viewContext)
		} catch {
			// TODO: Error handling
			print("*** ERROR: [\(#line)] \(#function) \(error)")
		}
    }
}

// MARK: - EventTimerControlDelegate
extension EventViewController {
    func startDateDidUpdate(_ startDate: Date!) {
		updateStartLabelWithDate(startDate)

		if let toDate = eventTimerControl?.nowDate {
			updateEventTimeFromDate(startDate, toDate: toDate)
		}
    }

    func nowDateDidUpdate(_ nowDate: Date!) {
		updateStopLabelWithDate(nowDate)

		if let fromDate = eventTimerControl?.startDate {
			updateEventTimeFromDate(fromDate, toDate: nowDate)
		}
	}

    func transformingDidUpdate(_ transform: EventTimerTransformingEnum) {
        switch transform {
        case .nowDateTransformingStart, .startDateTransformingStart:
            animateEventTransforming(transform)
        case .nowDateTransformingStop:
			guard let id = selectedEventID else {
				break
			}

			let result: Result<Event, FetchError> = fetch(url: id, inContext: persistentContainer.viewContext)

			guard let event = result.value() else {
				// For some reason there was no event
				// attached to this objectid, not at all what
				// was exspected
				// TODO: Error handling
				break
			}

            animateEventTransforming(transform)
            event.stopDate = eventTimerControl?.nowDate

			do {
				try save(context: persistentContainer.viewContext)
			} catch {
				// TODO: Error handling
				print("*** ERROR: [\(#line)] \(#function) \(error)")
			}
        case .startDateTransformingStop:
            animateEventTransforming(transform)
            if let startDate = eventTimerControl?.startDate,
				let id = selectedEventID
			{
				let result: Result<Event, FetchError> = fetch(url: id, inContext: persistentContainer.viewContext)

				guard let event = result.value() else {
					// For some reason there was no event
					// attached to this objectid, not at all what
					// was exspected
					// TODO: Error handling
					break
				}

                event.startDate = startDate

				do {
					try save(context: persistentContainer.viewContext)
				} catch {
					// TODO: Error handling
					print("*** ERROR: [\(#line)] \(#function) \(error)")
				}
			}
        default:
            break
        }
    }
}
